### ISSUE TITLE
Issues with DuckieTV on Ubuntu VMWARE

### ISSUE THREAD

**Issue by digi-ben** [SchizoDuckie/DuckieTV#1239]
I have a fresh ubnuntu build on a vmware virtual machine, its got the basics all sorted. qbittorent is already working fine, and VPN is too. I downloaded the latest nightly (and tried a few older ones too), but non of the copies of duckietv I have installed will open.

If you try the icon, it just does that click acknowledgement but nothing comes up. Cant see any processes running either.

If I try to run it in terminal with the command sudo DuckieTV, it just says "Opening in existing browser session." , and nothing else.

I thought maybe it wanted Chrome installing, tried that, no change.

Java isnt out to download anymore, I tried installing it from that tar command, but not sure if that one worked.

Any help would be greatly appreciated!

**Comment by garfield69** [comment]
strange.
I run an Ubuntu18 image under my win10 Oracle VM box.
1. I used the Chrome Browser to download the latest nightly from https://github.com/DuckieTV/Nightlies/releases/download/nightly-202001010130/DuckieTV-202001010130-ubuntu-x64.deb
2. I open the download folder and 
3. then I open the package with the right-mouse button and select the "open with software install" menu option 
4. I click the install button and enter my admin password
5. after the install is complete I close the installer and I right mouse click on the "activites" and search for duck
6. when the DuckieTV app icon shows up I click that to launch the app.

no problems.

If your app is starting and ending without opening a window then something is broken.
which duckietv nightly package did you install? could it be that you installed a x64 when  you need an x32?

or maybe you need to delete the DuckieTV working folder so that the app can start afresh with a clean slate.
you could try the following:
1. reboot your image to ensure there are no left over duckieTV processes or stubs
2. locate the duckieTV data folder by opening a terminal and `cd ~/.config/`
3. delete the `DuckieTV-Standalone` folder `sudo rm -r -f DuckieTV-Standalone`
4. locate the duckietv app icon and launch it

any luck?

**Comment by digi-ben** [comment]
Hiya, thanks for the reply!

Ok I have fully removed my copy of duckietv (it was the 64-bit nightly from the 1st Jan previously, on the 64bit Ubuntu I had). That includes the apt remove command, a purge and removing the folder as you said.

I have downloaded the 32bit nightly (DuckieTV-202001020130-ubuntu-x32.deb), ran the file which opens the ubuntu app installer. Hit install, asks for a password, and then it tries to go to 100% and it just stops and then the install button becomes available.
In terminal I can see there is no duckie folder either.
So I would guess it doesnt want a 32bit version.

So trying 64bit nightly (DuckieTV-202001020130-ubuntu-x64.deb), ran the file and its opening it in the ubuntu software installer. Hit Install, enter password. Its installed, I can see the folder in terminal. If I press the icon, it flashes, but no program launch.

From terminal I get this:
ben:~/.config/DuckieTV-Standalone$ sudo DuckieTV
[sudo] password for ben: 
Opening in existing browser session.

And nothing else happens. 

Could I be missing required software to make this work? Any help is seriously appreciated

**Comment by garfield69** [comment]
DuckieTV is written in [AngularJS](https://angularjs.org/) as a chrome browser extension.
The standalone app is the same JS code encapsulated in the [NWJS](https://nwjs.io/) run time environment, which uses the chromium engine to support running the chrome extension JS.

so if nwjs is failing to run then this is where we need to concentrate our investigation.

lets take Dtv and its standalone packages out of the equation and see if we can get nwjs to run.

according to http://docs.nwjs.io/en/latest/For%20Users/Getting%20Started/ you can perform a simple test by doing the following:

1. download the latest nwjs sdk 64-bit for linux from https://nwjs.io/downloads/
2. unpack the zip file
3. cd to the unpacked folder
4. start gedit or some other plain text editor and save `package.json` with the following content
```json
{
  "name": "helloworld",
  "main": "index.html"
}
```
5. start gedit and save `index.html` with the following content
```html
<!DOCTYPE html>
<html>
  <head>
    <title>Hello World!</title>
  </head>
  <body>
    <h1>Hello World!</h1>
  </body>
</html>
```
6. run nw

if it works you should see
![image](https://user-images.githubusercontent.com/6933240/71683165-a31ccd80-2df6-11ea-960f-9111e8a5117e.png)

**Comment by digi-ben** [comment]
Please excuse my ignorance, but how do I do the bit where you said run nw? Is there a particular deb I need to install?

**Comment by digi-ben** [comment]
Also, just out of curioristy, I tried to run Duckie again, and it came up saying my preferences could not be saved. Click OK. Nothing, but the process was now visible. 

So I ran the sudo DuckieTV command and got this...
_ben@:~/Downloads/nwjs-sdk-v0.43.3-linux-x64$ sudo DuckieTV
[5537:5563:0102/192712.160237:ERROR:bus.cc(394)] Failed to connect to the bus: Could not parse server address: Unknown address type (examples of valid types are "tcp" and on UNIX "unix")
[5585:5585:0102/192712.942208:FATAL:nw_content_renderer_hooks.cc(53)] Failed to load node library (error: libatomic.so.1: cannot open shared object file: No such file or directory)
#0 0x7fe178b0bc49 <unknown>

[5537:5548:0102/192815.133459:ERROR:ssl_client_socket_impl.cc(943)] handshake failed; returned -1, SSL error code 1, net_error -100
[5537:5548:0102/192815.136352:ERROR:ssl_client_socket_impl.cc(943)] handshake failed; returned -1, SSL error code 1, net_error -100
[5537:5548:0102/192825.174640:ERROR:ssl_client_socket_impl.cc(943)] handshake failed; returned -1, SSL error code 1, net_error -100
[5537:5548:0102/192825.188087:ERROR:ssl_client_socket_impl.cc(943)] handshake failed; returned -1, SSL error code 1, net_error -100
[5537:5548:0102/192844.317622:ERROR:ssl_client_socket_impl.cc(943)] handshake failed; returned -1, SSL error code 1, net_error -100
[5537:5548:0102/192848.047571:ERROR:ssl_client_socket_impl.cc(943)] handshake failed; returned -1, SSL error code 1, net_error -100_

Odd as nothing has really changed since the last time I tried, this box is purely for duckie and qbittorrent

**Comment by garfield69** [comment]
>Please excuse my ignorance, but how do I do the bit where you said run nw? Is there a particular deb I need to install?

from the terminal that is open in the nwjs folder that you upacked, you run nw by using the `./nw` command.

>So I ran the sudo DuckieTV command and got this..._ben@:~/Downloads/nwjs-sdk-v0.43.3-linux-x64$ sudo DuckieTV

ah no. the nwjs package you downloaded and are testing, is just the run time environment. it is the engine under which javascript chrome extensions can be run without having to use the chrome browser, hence the _standalone_ reference.
The nw environment is **not** duckietv, Dtv is just javascript and html, and the nwjs package you are testing does not contain any chrome extension apps (like dtv).
Does what I am saying make any sense to you?

when you run the DuckieTV app you are executing a bundled nwjs program with the Dtv javascript and html embedded in its code.
when you run the `./nw` you are running the raw nw app. the package.json and index.html that you created are in this instance, the raw equivalent of the Dtv javascript and html, that nw is going to load and execute.
I hope I have'nt confused you with an overload of information.

**Comment by digi-ben** [comment]
Thanks again!

So I ran the command and got..
_./nw
[6071:6071:0102/211809.847597:FATAL:nw_content_renderer_hooks.cc(53)] Failed to load node library (error: libatomic.so.1: cannot open shared object file: No such file or directory)
#0 0x7fb60c3bf059 <unknown>

[6073:6073:0102/211809.857435:FATAL:nw_content_renderer_hooks.cc(53)] Failed to load node library (error: libatomic.so.1: cannot open shared object file: No such file or directory)
#0 0x7f87b8b8a059 <unknown>_

So I ran **sudo apt-get install libatomic1**

Ran the ./nw command, and got the Hello World popup box!

So... I just ran **sudo DuckieTV** again, and got this:
![image](https://user-images.githubusercontent.com/59418209/71694125-337ef200-2da6-11ea-8d88-f053983a8a4f.png)

the error is constantly appearing in terminal, but if I close terminal it, it closes Duckie. I can run DuckieTV directly from the app menu now though too. No sign of errors.... 

It might be that libatomic1 was missing that caused the main issue, but I have noticed now that it will not save preferences at all, and there is an error running "context mismatch in svga_sampler_view_destroy" - I dont suppose you know if this is related?

**Comment by digi-ben** [comment]
Also, pushing my luck here. But I have had this error before when I had it running on a windows server. I managed to figure it out, but I didnt document how....

![image](https://user-images.githubusercontent.com/59418209/71694447-1dbdfc80-2da7-11ea-9bf2-c64532cbd5ce.png)

Do you happen to know the fix, this is on the latest nightly?

**Comment by garfield69** [comment]
so you are making progress. Excellent.
Launch the Nightly DuckieTV via the app icon, and not via the terminal.
Once DTV is running, use can use **F12** to open the *development Tools* panel and click on the **Console** tab so you can view the log and check that there are no significant issued with the javascript running.

re: qBt read https://github.com/SchizoDuckie/DuckieTV/wiki/Setting-up-qBitTorrent-3.2-with-DuckieTV and see if you can figure out what you need. If you cant work it out let me know.

**Comment by digi-ben** [comment]
Looks like quite a few errors:
![image](https://user-images.githubusercontent.com/59418209/71700222-2d474080-2dbb-11ea-9380-88a35c00f945.png)

The query error scrolls right the way down the page, it goes on for a while. Then it looks like it keeps trying the torrent connection on the default settings, which keep failing. 
Problem is, its not saving changes. So when I close and reopen duckietv, its back to defaults (tv shows, general settings and torrent settings). I remember that guide from last time, on your Ubuntu copy are you using qbittorrent? Is the CSRF an option still? On my WinServer I had that as an option, but its not there on this copy

Thank you again for all your help, its sincerely appreciated

**Comment by garfield69** [comment]
yes, the query error is a problem because it means that Dtv is unable to read/write to the database.
and the inability to save settings is probably part of the problem, because they are both kept in the same data folder.

so you are not out of the woods yet, something is still not right with your system requirements.

there is no point working on the qBt issue yet, not until we can get nwjs running properly, but for your reference: the CSRF checkbox that needs to be un-ticked is, on my linux qBt 4.2.1, labelled as
`Enable Cross-Site Request Forgery (CSRF) protection` on the qBt > Tools > Preferences > Web UI page.

so lets take a look at the database and environment settings.
start Dtv from the app icon, and at the first page, right click with the mouse and select the 
`inspect background page` menu option.
This will bring up the *Development Tools* page, but we are not looking at the foreground task where the Dtv html is processed, but the background task which is where the database should be available.

click on the **application** tab at the top, then expand the **Web SQL** on the left panel and click on the `seriesguide_chrome` sub set.
there should be 5 tables, Episodes, Fanart, Jackett, Seasons, Series.
do you have these?

then expand the **Local Storage** on the left panel, and click on the `Chrome extension...` table.
in the filters box type `database` and you should see
key value
database.version.Series 16
database.version.Seasons 6
database.version.Episodes 14
database.version.Fanart 1
database.version.Jackett 2

**Comment by digi-ben** [comment]
I'm not seeing that expansion of Web SQL as an option:
![image](https://user-images.githubusercontent.com/59418209/71701946-a303da00-2dc4-11ea-9c8b-391fcb9b5092.png)

Could that also be the issue?

**Comment by garfield69** [comment]
definitely. it explains why the foreground log was full of the query errors messages. 
so why is the websql database not been able to be created?

ok. lets go back to using the nw package that you managed to get the simple hello-world test working on. what we are going to do next is add the dtv source code to it and see if we can get something useful out of the terminal log.

so, shut down the standalone DuckieTV, and delete the working data folder and the test json and html.
1. `cd ~/.config/`
2. `sudo rm -r -f DuckieTV-Standalone`
3. `cd ~/Downloads/nwjs-sdk-v0.43.3-linux-x64` 
4. `rm package.json index.html`

now use your browser and download the latest Dtv Nightly source zip from https://github.com/DuckieTV/Nightlies/archive/nightly-202001020130.tar.gz
and unpack the contents to ~/Downloads/nwjs-sdk-v0.43.3-linux-x64

now run nw with `./nw` and this should run the Nightly DuckieTV if I've not forgotten anything.

hopefully the terminal log will show messages from nw complaining about websql or something useful.

**Comment by digi-ben** [comment]
Nothing in terminal at all, its very slow and sluggish and occasionally the screen will go the the "Aw Shucks" Google Chrome error page. But I was able to get into the area you asked about earlier...
![image](https://user-images.githubusercontent.com/59418209/71719360-04a06480-2e16-11ea-9f50-51db466d12c0.png)

Closed and opened it a few times using the ./nw command and its getting quite stable, and it remembers changes I make too. (That CSRF option really is missing in qbittorrent 4.0.3 - thats the latest I can install from the website or terminal, did you manually patch your version to get the option?)
![image](https://user-images.githubusercontent.com/59418209/71720429-9f4e7280-2e19-11ea-81b7-6ad71d36f85e.png)

**Comment by digi-ben** [comment]
Had to reboot the box after it froze up whilst checking the settings, I still had duckie on auto-start from before the removal steps you gave me earlier. So when the box restarted, DuckieTV fully loaded.... no errors!

I can also get into the database options, and there are no errors showing this time. Just noticed the folder in .config is back too... 

Do you think by deleting the config folder, and trying this debug stage, its somehow recreated the config folder and fixed the issue with the database having no read/write permissions?

Im just trying a database restore now from one I did on the WinServer box, but qbittorrent remains an issue. I was wondering if I could put the csrf option into a config file directly, or would that not make a difference as its version doesnt have the feature?

**Comment by garfield69** [comment]
if you are using qBt 4.0.3 then there is no CSRF option, that was introduced with versions 4.1 onward.
so you can forget about that.
now that you can save settings, have you set the Dtv torrent client to qBt 3.2+ interface, loaded that panel with your port 8087, the ip address (or localhost or 127.0.0.1) and username and password?

**Comment by garfield69** [comment]
I see in one of your recent posted pictures of the log, the CRUD is not defined error message.

if you are going to use the source version in combination with nw, on a machine that could be struggling with resources, then there is an additional step you need to do to improve reliability.
read and implement https://github.com/SchizoDuckie/DuckieTV/wiki/How-to-Install-a-Development-version-of-the-Windows-DuckieTV-Standalone-SOURCE#alert

**Comment by digi-ben** [comment]
Just made that change in your last comment thanks

Yes I have it set to Qbitorrent 3.2+
On the intergration I have tried it on localhost, 127.0.0.1, its lan IP. Same error, even with authentication settings changed. Although if I flip the port to 8080, I get a connect call failed.

I can load the WebUI fine from Chrome itself on the box, and from any other machine on the network too. Its frustrating as I remember the same issue on Windows Server, and I did something and got it to work but cant remember what :D

**Comment by garfield69** [comment]
what address do you use on your browser to access and login to the qBt webui from there?
what does the Dtv console log have when it tries to connect to your qBt 4.0.3?

**Comment by digi-ben** [comment]
Good call, so if I was to open the web interface, its just http://192.168.2.6:8087
And that works.

But I just spotted this in console:
![image](https://user-images.githubusercontent.com/59418209/71748862-c2097700-2e6b-11ea-9ed5-b2863eec4fde.png)

That address doesnt work in a web browser, but I imagine it needs it as part of a string?
![image](https://user-images.githubusercontent.com/59418209/71748915-ea917100-2e6b-11ea-8b93-2d469e9d35d8.png)

**Comment by garfield69** [comment]
so did you fill your Dtv qBt 3.2+ interface page like this
![image](https://user-images.githubusercontent.com/6933240/71750325-ef423d80-2edc-11ea-9ef1-6e321cd9cc25.png)
because the details are only **saved** if the connection is successful. 
if you did not get a connection and you restarted Dtv then what was there previously (from a backup import?) will remain.

**Comment by digi-ben** [comment]
Yes to filling up the interface, but as that error keeps coming up, it does indeed revert back to the generic settings on its next boot.
![image](https://user-images.githubusercontent.com/59418209/71751356-398ed480-2e73-11ea-9983-b8eb19180edf.png)

Even after I did a database restore, it didnt pull any settings with it. I thought a backup only remembered the tv shows?

![image](https://user-images.githubusercontent.com/59418209/71751404-56c3a300-2e73-11ea-8bf9-e962d1490571.png)
![image](https://user-images.githubusercontent.com/59418209/71751562-e23d3400-2e73-11ea-9661-cf9aef5041e6.png)

I have it setup to allow localhost through without a password, so I have tried it without authentication ticked, but still has the same message
![image](https://user-images.githubusercontent.com/59418209/71751457-85417e00-2e73-11ea-9820-5dfd685eb068.png)

**Comment by garfield69** [comment]
a backup restores settings, series and episodes.
after a backup import is competed I usually recommend a Dtv restart to be sure all the settings have been reloaded.

anytime you make changes to the qBt app settings, I recommend shutting it down via the menu exit and starting it up afresh. the qBt app is notorious for not refreshing things. (make sure have really closed the app, not go to system tray, or minimsed. I use the app's EXIT/QUIT menu.
Then restart Dtv as well, because it does not like qBt recycles while its active.
IF you changed the Port on qBt then reboot the system, because the port is left open but unresponsive by qBt and this can trick Dtv as it thinks it's got a port it can use when its not actually working.
IF you have set the qBt bypass authentication, then don't forget to also untick the Dtv qBt 3.2+ interface user-authorisation checkbox.
Anytime you test the Dtv qBt interface, and is does not work, I recommend on the next try, that you fill in all the fields afresh. Do not assume that the password in Dtv is valid.

the 401 response means that the request has not been applied because it lacks valid authentication credentials for the target resource.
this suggests that the issue is with the username or password.

so to recap, when was the last time you rebooted the system? 
after reboot, start up qBt, set up the authorisation, refresh the username and password and save and exit.
start qbt up again.
then start dtv and enter the interface data afresh. see if that helps.

**Comment by digi-ben** [comment]
Honestly I restart the box all the time, especially when it freezes up. 
But just to be sure, I have just done a quick test, did a backup, wiped the database, restarted, and tried just the torrent integration on its own and with no password, it still had the exact same error. Just restoring the backup again, which should of had all the details except the ip would of been off by one digit from the server move.

Also tried a manual shutdown of qbittorrent, duckietv, made sure processes were closed. Did a reboot, went into DuckieTV, deleted all of the connection details and typed it in manually and sadly no change. Same 401 error.... 

Side note, have you ever seen the backup restore take longer than usual?
![image](https://user-images.githubusercontent.com/59418209/71755193-6fd35080-2e81-11ea-9be8-48b594eba69f.png)

**Comment by digi-ben** [comment]
Could this be related? Its coming up during the backup restore:
![image](https://user-images.githubusercontent.com/59418209/71755397-4d8e0280-2e82-11ea-886a-73b89126a280.png)

**Comment by garfield69** [comment]
missing fanart is common and not an issue.
the problem is that your vm box is getting a bit overloaded trying to present lots of spinning wheel graphics and such.
when you have launched your import, I could switch back to the calendar view by clicking on the duckie icon in the top left corner. that should help speed things up. you can always open the console if you want to track progress and see the import-completed log message.

the qBt issue is a bit of a mystery. I will load up my vm box with that linux version and see if I can get my Dtv to talk to it. At least that way we will know for certain if its an issue with just that version or something else.

**Comment by digi-ben** [comment]
Tried and left it running over night but looks like it wont restore the backup.

Do you think If I perhaps set it up fully on my mac, could I transfer the config files to ubuntu? Then it would just fully work, if so do you know where they are stored?

**Comment by digi-ben** [comment]
I think I might of sorted out the qbitorrent integration issue. I think DuckieTV cant handle the version that ubuntu had in its repo. I did a repo update on qbitorrent, got a later version, which had the option to remove CSRF. And tried again.

At first it had a connect failed, did the connection attempt twice. Still failed. Clicked out. clicked on the monitor, and it switched over to connected!

Big issue now is, I cant seem to get this backup to restore. Its just showing the yellow circles spinning, and nothing appearing in calender. And if I go to my list, its blank and loading... 
(One problem after another with this guy I bet your thinking!!!)
![image](https://user-images.githubusercontent.com/59418209/71767206-b49ecc00-2f01-11ea-8790-a3597368a28d.png)
![image](https://user-images.githubusercontent.com/59418209/71767241-034c6600-2f02-11ea-9d71-9eebdceca1b9.png)

I tried manually deleting the config folders, and choosing wipe database, but each time I go to restore the backup, its the same fault. I even put 4 CPUs to task on the VM to see if it would help. It did restore them the other day though?

**Comment by garfield69** [comment]
I'm glad you managed to install qBt 4.2.1 and get it working with Dtv.
I tried building qBt 4.0.3 but it fails to compile and I can figure out why so I'm giving that idea up.

are you starting DuckieTV or are you starting ./nw ?
the log shows messages coming out of deps.js which would only be coming out of the Nightly DuckieTV build.

but I thought you were now using the Nightly source code in combo with nwjs now.

**Comment by digi-ben** [comment]
Indeed! Thats one obstacle!

Earlier in our thread, whilst doing the NWJS test, Duckie started to work without needing the ./nw command. Was very odd.
I have now revisited the original issue of the installation, and I think I have figured out another discovery. The deb file is not unpacking. Its showing in the Ubuntu Applications, but the actual package is not going into the opt/DuckieTV/DuckieTV folder.

Have you encountered this before?

**Comment by garfield69** [comment]
yes. if the .deb will not install it usually means that some dependency is missing so the software installer will not complete.
see https://itsfoss.com/install-deb-files-ubuntu/ for details, including two methods on how to install a .deb that required missing dependencies.

**Comment by digi-ben** [comment]
Any idea what the dependency was?

Ok so put this to rest, I have totally removed DuckieTV (to remove what we tried with nwjs etc). 
Tried reinstalling it with the regular installer but that will not work, so as you recommended I used that site and this command:

sudo dpkg -i (path to DuckieTV install deb)

It fully installs into the correct opt/ directory 👍 I might of had to edit the run path, but its loaded up and I have the fanart etc. I was able to restore the backup, took a little while but it worked this time and qbitorrent is now working too (now thats its updated!).
I'm not sure if having libatomic installed helped too? Or any of the other countless bits n bobs I tried :D 

I got to say, huge thankyou for all your help garfield69 - I never expected such amazing support, really impressed. Thank you!!
